File: src/GeneFromProtein.java
import java.util.ArrayList;

// This class is used to find genes in a DNA sequence.
public class GeneFromProtein {
    private final ArrayList<String> aminoAcidList = new ArrayList<>();
    private final ArrayList<String> geneList = new ArrayList<>();

    public ArrayList<String> getAminoAcid(String dna, String aminoAcid, ArrayList<String> isoleucine,
            ArrayList<String> leucine,
            ArrayList<String> valine, ArrayList<String> phenylalanine, ArrayList<String> methionine,
            ArrayList<String> cysteine, ArrayList<String> alanine, ArrayList<String> glycine, ArrayList<String> proline,
            ArrayList<String> threonine, ArrayList<String> serine, ArrayList<String> tyrosine,
            ArrayList<String> tryptophan, ArrayList<String> glutamine, ArrayList<String> asparagine,
            ArrayList<String> histidine, ArrayList<String> glutamicAcid, ArrayList<String> asparticAcid,
            ArrayList<String> lysine, ArrayList<String> arginine, ArrayList<String> stop) {

        // Maps the amino acid that the user entered to the start codon list.
        switch (aminoAcid) {
            case "isoleucine", "i", "ile" -> this.aminoAcidList.addAll(isoleucine);
            case "leucine", "l", "leu" -> this.aminoAcidList.addAll(leucine);
            case "valine", "v", "val" -> this.aminoAcidList.addAll(valine);
            case "phenylalanine", "f", "phe" -> this.aminoAcidList.addAll(phenylalanine);
            case "methionine", "m", "met" -> this.aminoAcidList.addAll(methionine);
            case "cysteine", "c", "cys" -> this.aminoAcidList.addAll(cysteine);
            case "alanine", "a", "ala" -> this.aminoAcidList.addAll(alanine);
            case "glycine", "g", "gly" -> this.aminoAcidList.addAll(glycine);
            case "proline", "p", "pro" -> this.aminoAcidList.addAll(proline);
            case "threonine", "t", "thr" -> this.aminoAcidList.addAll(threonine);
            case "serine", "s", "ser" -> this.aminoAcidList.addAll(serine);
            case "tyrosine", "y", "tyr" -> this.aminoAcidList.addAll(tyrosine);
            case "tryptophan", "w", "trp" -> this.aminoAcidList.addAll(tryptophan);
            case "glutamine", "q", "gln" -> this.aminoAcidList.addAll(glutamine);
            case "asparagine", "n", "asn" -> this.aminoAcidList.addAll(asparagine);
            case "histidine", "h", "his" -> this.aminoAcidList.addAll(histidine);
            case "glutamic acid", "e", "glu" -> this.aminoAcidList.addAll(glutamicAcid);
            case "aspartic acid", "d", "asp" -> this.aminoAcidList.addAll(asparticAcid);
            case "lysine", "k", "lys" -> this.aminoAcidList.addAll(lysine);
            case "arginine", "r", "arg" -> this.aminoAcidList.addAll(arginine);
            default -> System.out.println("Invalid amino acid");
        }
        for (String start_codon : this.aminoAcidList) {
            int start_index = dna.indexOf(start_codon.toLowerCase());
            for (String stop_codon : stop) {
                int stop_index = dna.indexOf(stop_codon.toLowerCase(), start_index + 3);
                if (start_index != -1 && stop_index != -1) {
                    this.geneList.add(dna.substring(start_index, stop_index + 3).toUpperCase());
                    break;
                }
            }
        }
        if (this.geneList.size() == 0) {
            for (int i = 0; i < 100; i++) {
                System.out.println();
            }
            this.geneList.add("No gene found");
        }
        return this.geneList;
    }
}

File: src/DNAnalyzer/Kotlin/GeneFromProtein.kt
package DNAnalyzer

// This class is used to find genes in a DNA sequence.
class GeneFromProtein {
    private val aminoAcidList: java.util.ArrayList<String> = java.util.ArrayList<String>()
    private val geneList: java.util.ArrayList<String> = java.util.ArrayList<String>()
    fun getAminoAcid(dna: String, aminoAcid: String?, isoleucine: java.util.ArrayList<String?>?,
                     leucine: java.util.ArrayList<String?>?,
                     valine: java.util.ArrayList<String?>?, phenylalanine: java.util.ArrayList<String?>?, methionine: java.util.ArrayList<String?>?,
                     cysteine: java.util.ArrayList<String?>?, alanine: java.util.ArrayList<String?>?, glycine: java.util.ArrayList<String?>?,
                     proline: java.util.ArrayList<String?>?,
                     threonine: java.util.ArrayList<String?>?, serine: java.util.ArrayList<String?>?, tyrosine: java.util.ArrayList<String?>?,
                     tryptophan: java.util.ArrayList<String?>?, glutamine: java.util.ArrayList<String?>?, asparagine: java.util.ArrayList<String?>?,
                     histidine: java.util.ArrayList<String?>?, glutamicAcid: java.util.ArrayList<String?>?,
                     asparticAcid: java.util.ArrayList<String?>?,
                     lysine: java.util.ArrayList<String?>?, arginine: java.util.ArrayList<String?>?, stop: java.util.ArrayList<String?>): java.util.ArrayList<String> {

        // Maps the amino acid that the user entered to the start codon list.
        when (aminoAcid) {
            "isoleucine", "i", "ile" -> aminoAcidList.addAll(isoleucine)
            "leucine", "l", "leu" -> aminoAcidList.addAll(leucine)
            "valine", "v", "val" -> aminoAcidList.addAll(valine)
            "phenylalanine", "f", "phe" -> aminoAcidList.addAll(phenylalanine)
            "methionine", "m", "met" -> aminoAcidList.addAll(methionine)
            "cysteine", "c", "cys" -> aminoAcidList.addAll(cysteine)
            "alanine", "a", "ala" -> aminoAcidList.addAll(alanine)
            "glycine", "g", "gly" -> aminoAcidList.addAll(glycine)
            "proline", "p", "pro" -> aminoAcidList.addAll(proline)
            "threonine", "t", "thr" -> aminoAcidList.addAll(threonine)
            "serine", "s", "ser" -> aminoAcidList.addAll(serine)
            "tyrosine", "y", "tyr" -> aminoAcidList.addAll(tyrosine)
            "tryptophan", "w", "trp" -> aminoAcidList.addAll(tryptophan)
            "glutamine", "q", "gln" -> aminoAcidList.addAll(glutamine)
            "asparagine", "n", "asn" -> aminoAcidList.addAll(asparagine)
            "histidine", "h", "his" -> aminoAcidList.addAll(histidine)
            "glutamic acid", "e", "glu" -> aminoAcidList.addAll(glutamicAcid)
            "aspartic acid", "d", "asp" -> aminoAcidList.addAll(asparticAcid)
            "lysine", "k", "lys" -> aminoAcidList.addAll(lysine)
            "arginine", "r", "arg" -> aminoAcidList.addAll(arginine)
            else -> println("Invalid amino acid")
        }
        for (start_codon in aminoAcidList) {
            val start_index: Int = dna.indexOf(start_codon.lowercase(Locale.getDefault()))
            for (stop_codon in stop) {
                val stop_index: Int = dna.indexOf(stop_codon.lowercase(Locale.getDefault()), start_index + 3)
                if (start_index != -1 && stop_index != -1) {
                    geneList.add(dna.substring(start_index, stop_index + 3).uppercase(Locale.getDefault()))
                    break
                }
            }
        }
        if (geneList.size == 0) {
            for (i in 0..99) {
                println()
            }
            geneList.add("No gene found")
        }
        return geneList
    }
}